web/text_layer_builder.js

```diff
@@ -83,37 +88,82 @@ class TextLayerBuilder {
    * @param {number} [timeout] - Wait for a specified amount of milliseconds
    *                             before rendering.
    */
-  render(timeout = 0) {
-    if (!(this.textContent || this.textContentStream) || this.renderingDone) {
+  render(viewport, timeout = 0) {
+    if (!(this.textContent || this.textContentStream)) {
       return;
     }
-    this.cancel();
 
-    this.textDivs.length = 0;
-    this.highlighter?.setTextMapping(this.textDivs, this.textContentItemsStr);
-    this.accessibilityManager?.setTextMapping(this.textDivs);
-
-    const textLayerFrag = document.createDocumentFragment();
-    this.textLayerRenderTask = renderTextLayer({
-      textContent: this.textContent,
-      textContentStream: this.textContentStream,
-      container: textLayerFrag,
-      viewport: this.viewport,
-      textDivs: this.textDivs,
-      textContentItemsStr: this.textContentItemsStr,
-      timeout,
-    });
-    this.textLayerRenderTask.promise.then(
-      () => {
-        this.textLayerDiv.append(textLayerFrag);
-        this.#finishRendering();
-        this.highlighter?.enable();
-        this.accessibilityManager?.enable();
-      },
-      function (reason) {
-        // Cancelled or failed to render text layer; skipping errors.
+    const scale = viewport.scale * (globalThis.devicePixelRatio || 1);
+
+    if (this.renderingDone) {
+      this.hide();
+      this.#update(viewport);
+      if (scale !== this.scale) {
+        updateTextLayer({
+          viewport,
+          scale,
+          textDivs: this.textDivs,
+          textDivProperties: this.textDivProperties,
+          isOffscreenCanvasSupported: this.isOffscreenCanvasSupported,
+        });
+        this.scale = scale;
       }
-    );
+      this.show();
+    } else {
+      this.cancel();
+
+      this.highlighter?.setTextMapping(this.textDivs, this.textContentItemsStr);
+      this.accessibilityManager?.setTextMapping(this.textDivs);
+
+      const textLayerFrag = document.createDocumentFragment();
+      this.textLayerRenderTask = renderTextLayer({
+        textContent: this.textContent,
+        textContentStream: this.textContentStream,
+        container: textLayerFrag,
+        viewport,
+        textDivs: this.textDivs,
+        textDivProperties: this.textDivProperties,
+        textContentItemsStr: this.textContentItemsStr,
+        isOffscreenCanvasSupported: this.isOffscreenCanvasSupported,
+        timeout,
+      });
+      this.textLayerRenderTask.promise.then(
+        () => {
+          this.div.append(textLayerFrag);
+          this.#finishRendering();
+          this.highlighter?.enable();
+          this.accessibilityManager?.enable();
```

Comment:
One thing that I'm not sure will work correctly here is highlighting of search results and the accessibility integration, since those may try and interact with the textLayer while it's "hidden".
Previously this obviously wasn't an issue, since the textLayer was always re-created, but it seems that it'll be problematical now.

The following is completely untested, and may require some tweaks, but could we address this simply by adding the following to the new `hide`-method:
```javascript
this.highlighter?.disable();
this.accessibilityManager?.disable();
```
and the following to the new `show`-method (rather than here):
```
this.highlighter?.enable();
this.accessibilityManager?.enable();
```

Reply:
In fact, this change leads to errors:
https://github.com/mozilla/pdf.js/blob/8fda3f04fe290a6ed7af0f684cdef7adea8cfcbe/web/text_accessibility.js#L117
really put the accessibility manager in a disabled state.
It's likely impossible to interact with the text layer when it's hidden but even if it was possible we can still access to its elements.
So I think it's fine, but of course if you can imagine something which shows I'm wrong, please tell me.

Reply:
Isn't that more of a problem specific to the `TextAccessibilityManager` implementation then, since it doesn't seem that the `TextHighlighter` is similarly affected?

The reason that I believe this to be a problem is that unless we disable the `TextHighlighter`, I really don't think that searching is going to work correctly unfortunately. Basically, we might end up trying to scroll invisible search results into view (or possibly other issues).
So, it seems to me that the `this.highlighter?.disable();`/`this.highlighter?.enable();` calls should stay in the `hide`/`show` method!?

Unfortunately, I don't know the `TextAccessibilityManager` well enough to tell what might happen if we leave it as-is...

Reply:
The only possibility to have something wrong with the accessibility manager is to create a new annotation when the text layer is invisible.
But in both cases, the text layer is hidden only when we update scale/rotation parameters, so I don't see how it's possible for a user to be able to change one of those parameters and then in be able to create an annotation or to make a search.
I just added a console.log just before the scrollMatchIntoView in text_highlighter.js, and with your proposed change I can see that this scroll... is called and it doesn't make sense for me to call it when zooming:  the scroll... must be only called when the user triggered a look for next/previous result.

Reply:
Please keep in mind that most things in the viewer is asynchronous, which means that the following order of events may now happen:
 1. The user triggers searching, which might take a while in a larger and more complex document since we need to fetch all textContent (which is async).
 2. The user *immediately* zooms the document, which "hides" the textLayer.
 3. The search operation finishes, and now tries to highlight/scroll *an invisible* match into view.
 4. The page finishes re-rendering, which "shows" the textLayer again.

I just don't see how the above is generally correct/safe, and it seems that we should still be disabling/enabling the highlighter just as before (when we re-created the textLayer).

> [...] is called and it doesn't make sense for me to call it when zooming: the scroll... must be only called when the user triggered a look for next/previous result.

Given the asynchronous nature of all of this, as outlined above, a search operation may be pending while zooming happens and in that case searching wouldn't actually work correctly.
Please note that the `PDFFindController.scrollMatchIntoView` method will only scroll *once*, to prevent the issues you mention: https://github.com/mozilla/pdf.js/blob/67741aeaa94956a8f75e756cafc7fcc51d520316/web/pdf_find_controller.js#L460-L473